docs/kb/config.md
---
order: 4
description: Primary configuration file to describe the development environment for your blockchain.
title: config.yml reference
---

# config.yml reference

The `config.yml` file generated in your blockchain folder uses key-value pairs to describe the development environment for your blockchain.

Only a default set of parameters is provided. If more nuanced configuration is required, you can add these parameters to the `config.yml` file.

## accounts

A list of user accounts created during genesis of the blockchain.

| Key      | Required | Type            | Description                                                                                                                     |
| -------- | -------- | --------------- | ------------------------------------------------------------------------------------------------------------------------------- |
| name     | Y        | String          | Local name of a key pair. An account name must be listed to gain access to the account tokens after the blockchain is launched. |
| coins    | Y        | List of Strings | Initial coins with denominations. For example, "1000token"                                                                      |
| address  | N        | String          | Account address in Bech32 address format.                                                                                        |
| mnemonic | N        | String          | Mnemonic used to generate an account. This field is ignored if `address` is specified.                                           |

**accounts example**

```yaml
accounts:
  - name: alice
    coins: ["1000token", "100000000stake"]
  - name: bob
    coins: ["500token"]
    address: cosmos1adn9gxjmrc3hrsdx5zpc9sj2ra7kgqkmphf8yw
```

## build

| Key      | Required | Type             | Description                                                                                                  |
|----------|----------|------------------|--------------------------------------------------------------------------------------------------------------|
| main     | N        | String           | When an app contains more than one main Go package, required to define the path of the chain's main package. |
| binary   | N        | String           | Name of the node binary that is built, typically ends with `d`.                                              |
| ldflags  | N        | List of Strings  | ldflags to set version information for go applications.                                                      |

**build example**

```yaml
build:
  binary: "mychaind"
  ldflags: [ "-X main.Version=development", "-X main.Date=01/05/2022T19:54" ]
```

### build.proto

| Key               | Required | Type            | Description                                                                                |
| ----------------- | -------- | --------------- | ------------------------------------------------------------------------------------------ |
| path              | N        | String          | Path to protocol buffer files. Default: `"proto"`.                                         |
| third_party_paths | N        | List of Strings | Path to third-party protocol buffer files. Default: `["third_party/proto", "proto_vendor"]`. |

## client

Configures and enables client code generation. To prevent Ignite CLI from regenerating the client, remove the `client` property.

### client.vuex

```yaml
client:
  vuex:
    path: "vue/src/store"
```

Generates TypeScript Vuex client for the blockchain in `path` on `serve` and `build` commands.

### client.openapi

```yaml
client:
  openapi:
    path: "docs/static/openapi.yml"
```

Generates OpenAPI YAML file in `path`. By default this file is embedded in the node's binary.

## faucet

The faucet service sends tokens to addresses. The default address for the web user interface is <http://localhost:4500>.

| Key               | Required | Type            | Description                                                 |
| ----------------- | -------- | --------------- | ----------------------------------------------------------- |
| name              | Y        | String          | Name of a key pair. The `name` key pair must be in `accounts`.            |
| coins             | Y        | List of Strings | One or more coins with denominations sent per request.       |
| coins_max         | N        | List of Strings | One or more maximum amounts of tokens sent for each address. |
| host              | N        | String          | Host and port number. Default: `:4500`                      |
| rate_limit_window | N        | String          | Time after which the token limit is reset (in seconds).      |

**faucet example**

```yaml
faucet:
  name: faucet
  coins: ["100token", "5foo"]
  coins_max: ["2000token", "1000foo"]
  port: 4500
```

## validator

A blockchain requires one or more validators.

| Key    | Required | Type   | Description                                                                                     |
| ------ | -------- | ------ | ----------------------------------------------------------------------------------------------- |
| name   | Y        | String | The account that is used to initialize the validator. The `name` key pair must be in `accounts`. |
| staked | Y        | String | Amount of coins to bond. Must be less than or equal to the amount of coins in the account.       |

**validator example**

```yaml
accounts:
  - name: alice
    coins: ["1000token", "100000000stake"]
validator:
  name: user1
  staked: "100000000stake"
```

## init.home

The path to the data directory that stores blockchain data and blockchain configuration.

**init example**

```yaml
init:
  home: "~/.myblockchain"
```

## init.config

Overwrites properties in `config/config.toml` in the data directory.

## init.app

Overwrites properties in `config/app.toml` in the data directory.

## init.client

Overwrites properties in `config/client.toml` in the data directory.

**init.client example**

```yaml
init:
  client:
    keyring-backend: "os"
```

## host

Configuration of host names and ports for processes started by Ignite CLI:

**host example**

```yaml
host:
  rpc: ":26659"
  p2p: ":26658"
  prof: ":6061"
  grpc: ":9091"
  api: ":1318"
```

## genesis

Use to overwrite values in `genesis.json` in the data directory to test different values in development environments. See [Genesis overwrites for development](../kb/genesis.md).


changelog.md
# Changelog

## [`v0.20.4`](https://github.com/ignite-hq/cli/releases/tag/v0.20.4)

### Fixes

- Use `protoc` binary compiled in an older version of macOS AMD64 for backwards compatibility in code generation


## [`v0.20.3`](https://github.com/ignite-hq/cli/releases/tag/v0.20.3)

### Fixes

- Use latest version of CLI in templates to fix Linux ARM support _(It's now possible to develop chains in Linux ARM machines and since the chain depends on the CLI in its go.mod, it needs to use the latest version that support ARM targets)_

## [`v0.20.2`](https://github.com/ignite-hq/cli/releases/tag/v0.20.2)

### Fixes

- Use `unsafe-reset-all` cmd under `tendermint` cmd for chains that use `=> v0.45.3` version of Cosmos SDK

## [`v0.20.1`](https://github.com/ignite-hq/cli/releases/tag/v0.20.1)

### Features

- Release the CLI with Linux ARM and native M1 binaries

## [`v0.20.0`](https://github.com/ignite-hq/cli/releases/tag/v0.20.0)

Our new name is **Ignite CLI**!

**IMPORTANT!** This upgrade renames `starport` command to `ignite`. From now on, use `ignite` command to access the CLI.

### Features

- Upgraded Cosmos SDK version to `v0.45.2`
- Added support for in memory backend in `pkg/cosmosclient` package
- Improved our tutorials and documentation

## [`v0.19.5`](https://github.com/ignite-hq/cli/pull/2158/commits)

### Features

- Enable client code and Vuex code generation for query only modules as well.
- Upgraded the Vue template to `v0.3.5`.

### Fixes:
- Fixed snake case in code generation.
- Fixed plugin installations for Go =>v1.18.

### Changes:
- Dropped transpilation of TS to JS. Code generation now only produces TS files.

## `v0.19.4`

### Features

- Upgraded Vue template to `v0.3.0`.

## `v0.19.3`

### Features

- Upgraded Flutter template to `v2.0.3`

## [`v0.19.2`](https://github.com/ignite-hq/cli/milestone/14)

### Fixes

- Fixed race condition during faucet transfer
- Fixed account sequence mismatch issue on faucet and relayer
- Fixed templates for IBC code scaffolding

### Features

- Upgraded blockchain templates to use IBC v2.0.2

### Breaking Changes

- Deprecated the Starport Modules [tendermint/spm](https://github.com/tendermint/spm) repo and moved the contents to the Ignite CLI repo [`ignite/pkg/`](https://github.com/ignite-hq/cli/tree/develop/ignite/pkg/) in [PR 1971](https://github.com/ignite-hq/cli/pull/1971/files) 
 
    Updates are required if your chain uses these packages: 

    - `spm/ibckeeper` is now `pkg/cosmosibckeeper`
    - `spm/cosmoscmd` is now `pkg/cosmoscmd` 
    - `spm/openapiconsole` is now `pkg/openapiconsole`
    - `testutil/sample` is now `cosmostestutil/sample`

- Updated the faucet HTTP API schema. See API changes in [fix: improve faucet reliability #1974](https://github.com/ignite-hq/cli/pull/1974/files#diff-0e157f4f60d6fbd95e695764df176c8978d85f1df61475fbfa30edef62fe35cd)

## `v0.19.1`

### Fixes

- Enabled the `scaffold flutter` command

## `v0.19.0`

### Features

- `starport scaffold` commands support `ints`, `uints`, `strings`, `coin`, `coins` as field types (#1579)
- Added simulation testing with `simapp` to the default template (#1731)
- Added `starport generate dart` to generate a Dart client from protocol buffer files
- Added `starport scaffold flutter` to scaffold a Flutter mobile app template
- Parameters can be specified with a new `--params` flag when scaffolding modules (#1716)
- Simulations can be run with `starport chain simulate`
- Set `cointype` for accounts in  `config.yml` (#1663)

### Fixes

- Allow using a `creator` field when scaffolding a model with a `--no-message` flag (#1730)
- Improved error handling when generating code (#1907)
- Ensure account has funds after faucet transfer when using `cosmosclient` (#1846)
- Move from `io/ioutil` to `io` and `os` package (refactoring) (#1746)

## `v0.18.0`

### Breaking Changes

- Starport v0.18 comes with Cosmos SDK v0.44 that introduced changes that are not compatible with chains that were scaffolded with Starport versions lower than v0.18. After upgrading from Starport v0.17.3 to Starport v0.18, you must update the default blockchain template to use blockchains that were scaffolded with earlier versions. See [Migration](./docs/migration/index.md).

### Features:

- Scaffold commands allow using previously scaffolded types as fields
- Added `--signer` flag to `message`, `list`, `map`, and `single` scaffolding to allow customizing the name of the signer of the message
- Added `--index` flag to `scaffold map` to provide a custom list of indices
- Added `scaffold type` to scaffold a protocol buffer definition of a type
- Automatically check for new Starport versions
- Added `starport tools completions` to generate CLI completions
- Added `starport account` commands to manage accounts (key pairs)
- `starport version` now prints detailed information about OS, Go version, and more
- Modules are scaffolded with genesis validation tests
- Types are scaffolded with tests for `ValidateBasic` methods
- `cosmosclient` has been refactored and can be used as a library for interacting with Cosmos SDK chains
- `starport relayer` uses `starport account`
- Added `--path` flag for all `scaffold`, `generate` and `chain` commands
- Added `--output` flag to the `build` command
- Configure port of gRPC web in `config.yml` with the `host.grpc-web` property
- Added `build.main` field to `config.yml` for apps to specify the path of the chain's main package. This property is required to be set only when an app contains multiple main packages.

### Fixes

- Scaffolding a message no longer prevents scaffolding a map, list, or single that has the same type name when using the `--no-message` flag
- Generate Go code from proto files only from default directories or directories specified in `config.yml`
- Fixed faucet token transfer calculation
- Removed `creator` field for types scaffolded with the `--no-message` flag
- Encode the count value in the store with `BigEndian`

## `v0.17.3`

### Fixes


ignite/chainconfig/config.go
package chainconfig

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"

	"github.com/goccy/go-yaml"
	"github.com/imdario/mergo"

	"github.com/ignite-hq/cli/ignite/pkg/xfilepath"
)

var (
	// ConfigDirPath returns the path of configuration directory of Ignite.
	ConfigDirPath = xfilepath.JoinFromHome(xfilepath.Path(".ignite"))

	// ConfigFileNames is a list of recognized names as for Ignite's config file.
	ConfigFileNames = []string{"config.yml", "config.yaml"}
)

var (
	// ErrCouldntLocateConfig returned when config.yml cannot be found in the source code.
	ErrCouldntLocateConfig = errors.New(
		"could not locate a config.yml in your chain. please follow the link for" +
			"how-to: https://github.com/ignite-hq/cli/blob/develop/docs/configure/index.md")
)

// DefaultConf holds default configuration.
var DefaultConf = Config{
	Host: Host{
		// when in Docker on MacOS, it only works with 0.0.0.0.
		RPC:     "0.0.0.0:26657",
		P2P:     "0.0.0.0:26656",
		Prof:    "0.0.0.0:6060",
		GRPC:    "0.0.0.0:9090",
		GRPCWeb: "0.0.0.0:9091",
		API:     "0.0.0.0:1317",
	},
	Build: Build{
		Proto: Proto{
			Path: "proto",
			ThirdPartyPaths: []string{
				"third_party/proto",
				"proto_vendor",
			},
		},
	},
	Faucet: Faucet{
		Host: "0.0.0.0:4500",
	},
}

// Config is the user given configuration to do additional setup
// during serve.
type Config struct {
	Accounts  []Account              `yaml:"accounts"`
	Validator Validator              `yaml:"validator"`
	Faucet    Faucet                 `yaml:"faucet"`
	Client    Client                 `yaml:"client"`
	Build     Build                  `yaml:"build"`
	Init      Init                   `yaml:"init"`
	Genesis   map[string]interface{} `yaml:"genesis"`
	Host      Host                   `yaml:"host"`
}

// AccountByName finds account by name.
func (c Config) AccountByName(name string) (acc Account, found bool) {
	for _, acc := range c.Accounts {
		if acc.Name == name {
			return acc, true
		}
	}
	return Account{}, false
}

// Account holds the options related to setting up Cosmos wallets.
type Account struct {
	Name     string   `yaml:"name"`
	Coins    []string `yaml:"coins,omitempty"`
	Mnemonic string   `yaml:"mnemonic,omitempty"`
	Address  string   `yaml:"address,omitempty"`
	CoinType string   `yaml:"cointype,omitempty"`

	// The RPCAddress off the chain that account is issued at.
	RPCAddress string `yaml:"rpc_address,omitempty"`
}

// Validator holds info related to validator settings.
type Validator struct {
	Name   string `yaml:"name"`
	Staked string `yaml:"staked"`
}

// Build holds build configs.
type Build struct {
	Main    string   `yaml:"main"`
	Binary  string   `yaml:"binary"`
	LDFlags []string `yaml:"ldflags"`
	Proto   Proto    `yaml:"proto"`
}

// Proto holds proto build configs.
type Proto struct {
	// Path is the relative path of where app's proto files are located at.
	Path string `yaml:"path"`

	// ThirdPartyPath is the relative path of where the third party proto files are
	// located that used by the app.
	ThirdPartyPaths []string `yaml:"third_party_paths"`
}

// Client configures code generation for clients.
type Client struct {
	// TSClient configures code generation for Typescript Client.
	Typescript Typescript `yaml:"typescript"`

	// Dart configures client code generation for Dart.
	Dart Dart `yaml:"dart"`

	// OpenAPI configures OpenAPI spec generation for API.
	OpenAPI OpenAPI `yaml:"openapi"`
}

// TSClient configures code generation for Typescript Client.
type Typescript struct {
	// Path configures out location for generated Typescript Client code.
	Path string `yaml:"path"`
}

// Dart configures client code generation for Dart.
type Dart struct {
	// Path configures out location for generated Dart code.
	Path string `yaml:"path"`
}

// OpenAPI configures OpenAPI spec generation for API.
type OpenAPI struct {
	Path string `yaml:"path"`
}

// Faucet configuration.
type Faucet struct {
	// Name is faucet account's name.
	Name *string `yaml:"name"`

	// Coins holds type of coin denoms and amounts to distribute.
	Coins []string `yaml:"coins"`

	// CoinsMax holds of chain denoms and their max amounts that can be transferred
	// to single user.
	CoinsMax []string `yaml:"coins_max"`

	// LimitRefreshTime sets the timeframe at the end of which the limit will be refreshed
	RateLimitWindow string `yaml:"rate_limit_window"`

	// Host is the host of the faucet server
	Host string `yaml:"host"`

	// Port number for faucet server to listen at.
	Port int `yaml:"port"`
}

// Init overwrites sdk configurations with given values.
type Init struct {
	// App overwrites appd's config/app.toml configs.
	App map[string]interface{} `yaml:"app"`

	// Client overwrites appd's config/client.toml configs.
	Client map[string]interface{} `yaml:"client"`

	// Config overwrites appd's config/config.toml configs.
	Config map[string]interface{} `yaml:"config"`

	// Home overwrites default home directory used for the app
	Home string `yaml:"home"`

	// KeyringBackend is the default keyring backend to use for blockchain initialization
	KeyringBackend string `yaml:"keyring-backend"`
}

// Host keeps configuration related to started servers.
type Host struct {
	RPC     string `yaml:"rpc"`
	P2P     string `yaml:"p2p"`
	Prof    string `yaml:"prof"`
	GRPC    string `yaml:"grpc"`
	GRPCWeb string `yaml:"grpc-web"`
	API     string `yaml:"api"`
}

// Parse parses config.yml into UserConfig.
func Parse(r io.Reader) (Config, error) {
	var conf Config
	if err := yaml.NewDecoder(r).Decode(&conf); err != nil {
		return conf, err
	}
	if err := mergo.Merge(&conf, DefaultConf); err != nil {
		return Config{}, err
	}
	return conf, validate(conf)
}

// ParseFile parses config.yml from the path.
func ParseFile(path string) (Config, error) {
	file, err := os.Open(path)
	if err != nil {
		return Config{}, nil
	}
	defer file.Close()
	return Parse(file)
}

// validate validates user config.
func validate(conf Config) error {
	if len(conf.Accounts) == 0 {
		return &ValidationError{"at least 1 account is needed"}
	}
	if conf.Validator.Name == "" {
		return &ValidationError{"validator is required"}
	}
	return nil
}

// ValidationError is returned when a configuration is invalid.
type ValidationError struct {
	Message string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("config is not valid: %s", e.Message)
}

// LocateDefault locates the default path for the config file, if no file found returns ErrCouldntLocateConfig.
func LocateDefault(root string) (path string, err error) {
	for _, name := range ConfigFileNames {
		path = filepath.Join(root, name)
		if _, err := os.Stat(path); err == nil {
			return path, nil
		} else if !os.IsNotExist(err) {
			return "", err
		}
	}
	return "", ErrCouldntLocateConfig
}

// FaucetHost returns the faucet host to use
func FaucetHost(conf Config) string {
	// We keep supporting Port option for backward compatibility
	// TODO: drop this option in the future
	host := conf.Faucet.Host
	if conf.Faucet.Port != 0 {
		host = fmt.Sprintf(":%d", conf.Faucet.Port)
	}

	return host
}

// CreateConfigDir creates config directory if it is not created yet.
func CreateConfigDir() error {
	confPath, err := ConfigDirPath()
	if err != nil {
		return err
	}

	return os.MkdirAll(confPath, 0755)
}
